### ISSUE TITLE
Shot queueing mechanics

### ISSUE THREAD

**Issue by kaytotes** [Project-Epoch/bugtracker#8327]
[Hunter] Shot queueing mechanics
Details
The current way in which Steady Shot works overrides the auto shot timer which doesn't allow for the basic mechanics of the class to work (1:1 and 3:2 rotations. )

How it should work
1. Auto shot is a spell with a 0.5s cast time (affected by hast effect) and its cooldown is the hunter's ranged attack speed minus 0.5
2. Instant cast spells don't interact with auto shot in any way.
3. Auto shot can only be cast if you're not currently casting any spell and only while standing still.
4. Casting any spell with a cast time or moving should interrupt the 0.5s auto shot cast time.
5. Steady shot and multi-shot are the only exceptions to rule number 4, both spells don't interrupt auto
shot.
6. If you cast steady shot or multi-shot while an auto shot is mid cast, your next steady shot or multi-shot
should start casting as soon as the 0.5s auto shot cast time finishes.
7. Whenever you finish casting an aimed shot, your auto shot cooldown is reset.

Now this next part is crucial to understand why the 3:2 rotation works the way it does (called GCD desyncing):

8. If you queue a steady shot or multi-shot during the auto shot wind up time, your GCD starts when the game client send the spell cast to the server.
9. Steady shot has a cast time shorter than a GCD, since ranged haste doesn't affect GCD.
10. Spell queuing should respect GCD limitations, meaning that you can't cast 2 spells back to back if both of them have a cast time shorter than a GCD.
11. If you cast a steady shot during the auto shot wind up time, the spell triggers a GCD before the spell cast begins, allowing you to cast 2 back to back steady shots.
12. You can't cast 2 back to back steady shots if auto shot is off cooldown by the time the first steady shot cast finishes. Auto shot should always take precedence over steady/multi shot.

More details here : https://github.com/SunstriderEmu/BugTracker/issues/224
GCD Desync explaination : 
https://boukx.github.io/threetwo/
video (from Endless.gg server, Stormforge server is using it also): 
https://streamable.com/ow5e1e

**Comment by kaytotes** [comment]
Note. Design decision to be made whether we consider Classic or OG TBC as "the source of truth"

**Comment by kaytotes** [comment]
tl;dr We decided that TBC Classic is our benchmark not OG TBC 2.4.3 which means that we won't be implementing this as it did not function in Classic and seems to have been unintended behaviour.

```
GCD desyncing currently does not work in classic. When you queue a spell, the GCD for that spell begins when the spell begins casting.

We’re not confident why GCD desyncing worked in TBC, but if I had to hazard a wild guess I’d say it was an inconsistency brought about by some part of GCDs being handled client client sided while casts/actions were being verified more strictly on the server side. I bring this up to say that I doubt GCD desyncing will be re-implemented for TBC since its such a weird behavior touching both server and client side handling of spell casting which only effects hunters (you need a core rotation of spells with cast times less than the GCD and casts which do not have GCDs to benefit from the tech).
```

**Comment by BinBreton** [comment]
https://web.archive.org/web/20081014175529/http://forums.wow-europe.com/thread.html?topicId=2062236264&pageNo=1&sid=1

Shot queueing was implemented shortly after this post.
